File: src/stores/collections.ts
import { createSignal } from "solid-js";
import { createStore, produce } from "solid-js/store";
import * as api from "../lib/api";
import { triggerPush } from "../lib/sync";
import { scheduleSave } from "../lib/session";

export interface CollectionNode {
  collection: api.Collection;
  children: CollectionNode[];
  requests: api.SavedRequest[];
  expanded: boolean;
}

const [teams, setTeams] = createSignal<api.Team[]>([]);
const [activeTeam, setActiveTeam] = createSignal<string>("");
const [collections, setCollections] = createStore<CollectionNode[]>([]);
const [loading, setLoading] = createSignal(false);

// Alias for backward compatibility
const activeWorkspace = activeTeam;

export { teams, activeTeam, activeWorkspace, collections, loading };

// --- Last used collection tracking ---
const LAST_COLLECTION_KEY = "last_used_collection_id";
const [lastUsedCollectionId, setLastUsedCollectionIdRaw] = createSignal<string | null>(
  localStorage.getItem(LAST_COLLECTION_KEY)
);

export function setLastUsedCollectionId(id: string) {
  setLastUsedCollectionIdRaw(id);
  localStorage.setItem(LAST_COLLECTION_KEY, id);
}

export function getDefaultCollectionId(): string | null {
  const last = lastUsedCollectionId();
  if (last) return last;
  if (collections.length > 0) return collections[0].collection.id;
  return null;
}

// --- Folder expand state ---
const [expandedFolders, setExpandedFolders] = createSignal<Set<string>>(new Set<string>());
export { expandedFolders, setExpandedFolders };

export function expandFolder(id: string) {
  setExpandedFolders(prev => {
    if (prev.has(id)) return prev;
    const next = new Set(prev);
    next.add(id);
    return next;
  });
  scheduleSave();
}

export function toggleFolder(id: string) {
  setExpandedFolders(prev => {
    const next = new Set(prev);
    if (next.has(id)) next.delete(id);
    else next.add(id);
    return next;
  });
  scheduleSave();
}

const [refreshTrigger, setRefreshTrigger] = createSignal(0);
export { refreshTrigger };
export function triggerRefresh() {
  setRefreshTrigger(refreshTrigger() + 1);
  const teamId = activeTeam();
  if (teamId) loadCollections(teamId);
}

export async function loadTeams() {
  const ts = await api.listTeams();
  setTeams(ts);
  if (ts.length > 0 && !activeTeam()) {
    setActiveTeam(ts[0].id);
    await loadCollections(ts[0].id);
  }
}

export async function loadCollections(teamId: string) {
  setLoading(true);
  try {
    const cols = await api.listCollections(teamId);
    const nodes: CollectionNode[] = [];
    const map = new Map<string, CollectionNode>();

    for (const col of cols) {
      const node: CollectionNode = { collection: col, children: [], requests: [], expanded: false };
      map.set(col.id, node);
    }

    for (const col of cols) {
      const node = map.get(col.id)!;
      if (col.parent_id && map.has(col.parent_id)) {
        map.get(col.parent_id)!.children.push(node);
      } else {
        nodes.push(node);
      }
    }

    // Load requests for each collection
    for (const col of cols) {
      const reqs = await api.listRequests(col.id);
      const node = map.get(col.id)!;
      node.requests = reqs;
    }

    setCollections(nodes);
  } finally {
    setLoading(false);
  }
}

export async function addCollection(name: string, parentId: string | null = null) {
  const teamId = activeTeam();
  if (!teamId) return;
  await api.createCollection(teamId, parentId, name);
  await loadCollections(teamId);
  triggerPush();
}

export async function removeCollection(id: string) {
  const teamId = activeTeam();
  if (!teamId) return;

  // Rust backend handles recursive deletion of children + tombstone creation
  await api.deleteCollection(id);
  await loadCollections(teamId);
  triggerPush();
}

export async function addRequest(collectionId: string, name: string, method: string = "GET") {
  setLastUsedCollectionId(collectionId);
  await api.createRequest(collectionId, name, method, "");
  const teamId = activeTeam();
  if (teamId) await loadCollections(teamId);
  triggerPush();
}

export async function removeRequest(id: string) {
  try {
    await api.deleteRequest(id);
  } catch (err) {
    console.error("Failed to delete request:", err);
  }
  const teamId = activeTeam();
  if (teamId) await loadCollections(teamId);
  triggerPush();
}

export async function moveRequest(id: string, targetCollectionId: string) {
  try {
    await api.moveRequest(id, targetCollectionId);
    setLastUsedCollectionId(targetCollectionId);
    const teamId = activeTeam();
    if (teamId) await loadCollections(teamId);
    triggerPush();
  } catch (err) {
    console.error("Failed to move request:", err);
  }
}

File: src-tauri/src/commands/collections.rs
use crate::state::AppState;
use tenso_shared::models::*;
use std::sync::Arc;

#[tauri::command]
pub async fn list_teams(state: tauri::State<'_, Arc<AppState>>) -> Result<Vec<Team>, String> {
    state.db.list_teams().map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn create_team(state: tauri::State<'_, Arc<AppState>>, name: String) -> Result<Team, String> {
    state.db.create_team(&name).map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn list_collections(state: tauri::State<'_, Arc<AppState>>, team_id: String) -> Result<Vec<Collection>, String> {
    state.db.list_collections(&team_id).map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn create_collection(state: tauri::State<'_, Arc<AppState>>, team_id: String, parent_id: Option<String>, name: String) -> Result<Collection, String> {
    state.db.create_collection(&team_id, parent_id.as_deref(), &name).map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn update_collection(state: tauri::State<'_, Arc<AppState>>, id: String, name: String) -> Result<(), String> {
    state.db.update_collection(&id, &name).map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn delete_collection(state: tauri::State<'_, Arc<AppState>>, id: String) -> Result<(), String> {
    state.db.delete_collection(&id).map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn list_requests(state: tauri::State<'_, Arc<AppState>>, collection_id: String) -> Result<Vec<SavedRequest>, String> {
    state.db.list_requests(&collection_id).map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn get_request(state: tauri::State<'_, Arc<AppState>>, id: String) -> Result<Option<SavedRequest>, String> {
    state.db.get_request(&id).map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn create_request(state: tauri::State<'_, Arc<AppState>>, collection_id: String, name: String, method: String, url: String) -> Result<SavedRequest, String> {
    state.db.create_request(&collection_id, &name, &method, &url).map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn update_request(state: tauri::State<'_, Arc<AppState>>, request: SavedRequest) -> Result<(), String> {
    state.db.update_request(&request).map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn delete_request(state: tauri::State<'_, Arc<AppState>>, id: String) -> Result<(), String> {
    state.db.delete_request(&id).map_err(|e| e.to_string())
}
